### PR TITLE
Autograd for ATenOp

### PR DESCRIPTION
**Pull request by iK1D**
Add autograd support for ATenOp.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc [hunk 1]
```diff
@@ -159,41 +160,135 @@ class ATenOperatorCache {
   std::unordered_map<std::pair<std::string, std::string>, ATenOperator, PairHash> ops_;
 };
 
+// AutogradContext saves forward inputs that require grad, and the output tensor.
```
Comment (wschin) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:163]: Very smart idea!

**Comment by SherlockNoMad** [comment]
If the forward node has more than one outputs, and they could be needed for backward computation. 


For example: 
```
        NodeDef(OpDef{"LayerNormalizationGrad", kMSDomain, 1},
                {GO(0), I(0), I(1), **O(1), O(2)**},
                {GI(0), GI(1), GI(2)},
                {SrcNodeAttributes()})};
```

```
- name: native_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor)
  input, weight, bias: "grad.defined() ? native_batch_norm_backward(grad, input, weight, running_mean, running_var, **result1, result2,** training, eps, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
```

**Comment by iK1D** [comment]
> If the forward node has more than one outputs, and they could be needed for backward computation for some op, e.g. LayerNorm and BacthNorm.
> 
> As we are not saving output IValues to AutogradContext, can pytorch autograd engine handling such cases?
> 
> For example:
> 
> ```
>         NodeDef(OpDef{"LayerNormalizationGrad", kMSDomain, 1},
>                 {GO(0), I(0), I(1), **O(1), O(2)**},
>                 {GI(0), GI(1), GI(2)},
>                 {SrcNodeAttributes()})};
> ```
> 
> ```
> - name: native_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor)
>   input, weight, bias: "grad.defined() ? native_batch_norm_backward(grad, input, weight, running_mean, running_var, **result1, result2,** training, eps, grad_input_mask) : std::tuple<Tensor, Tensor, Tensor>()"
> ```

It actually depends on which level we export the ATenOp. Normally it should be at the generic level, where LayerNorm, BatchNorm, Unfold will only return 1 tensor, and we will call backward from this tensor. Inside the implemtation, PyTorch will call different low level functions, which will return more than one outputs (even different number of outputs on different device). But PyTorch's autograd will handle this well internally. Take the Unfold as example, before we use autograd, we need to register the ATenOp to aten::max_pool2d_with_indices, which will return both output and indices, and pass indices tensor as gradient input. But when we use autograd, we can use aten::unfold directly, which return only one output. The indices tensor as an edge still exists, but it's inside autograd's logic.

I don't see any common PyTorch op that return more than one tensor as outputs for now, so here I save only the first tensor (maybe we can assert there is only one input). If there is ops that return more than one tensor, we can change our code to use torch::autograd::grad (https://pytorch.org/docs/stable/generated/torch.autograd.grad.html), which can take more than one outputs and inputs.

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 2]
```diff
@@ -2820,7 +2820,21 @@ Return true if all elements are true and false otherwise.
...
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetSupportLevel(OpSchema::SupportType::EXPERIMENTAL)
+      .SetDoc("ATenOpGrad")
+      .Input(0, "output_grad", "ATenOpGrad's output grad.", "T")
+      .Input(1, "context_id", "Context ID generated by the corresponding ATenOp.", "TInt64")
+      .Output(0, "input_grads", "ATenOpGrad's input grads.", "T", OpSchema::Variadic,
+              /*is_homogeneous*/ false,
+              /*min_arity*/ 1)
+      .TypeConstraint("T", OpSchema::all_tensor_types(), "Allow inputs and outputs to be any kind of tensor.")
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2836]: gradient can only be floats.

File: orttraining/orttraining/core/graph/gradient_builder_base.cc [hunk 3]
```diff
@@ -405,5 +405,22 @@ AttributeProto GradientBuilderBase::AttributeDefinitionToAttributeProto(
   return attr_proto;
 }
 
+// NOTE THAT we are changing the ATenOp forward node. If any of the input requires grad, we need to:
+// 1. set the requires_grad attribute in ATenOp so we can get this info during ATenOp forward execution and pass to
+// ATenOp executor.
+// 2. Append a new CPU INT scalar to outputs to pass context ID to ATenOpGrad for backward execution.
+void GradientBuilderBase::HandleATenOpGradient(const std::vector<int64_t>& requires_grad) const {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder_base.cc:412]: I feel  HandleATenOpGradient should be a private function for GetATenOpGradient class. 

we might skip using DECLARE_GRADIENT_BUILDER macro, and use an explicit GetATenOpGradient  class declaration?

File: orttraining/orttraining/core/graph/gradient_builder.h [hunk 4]
```diff
@@ -77,6 +77,7 @@ DECLARE_GRADIENT_BUILDER(GetPadGradient)
 DECLARE_GRADIENT_BUILDER(GetIdentityGradient)
 DECLARE_GRADIENT_BUILDER(GetPythonOpGradient)
 DECLARE_GRADIENT_BUILDER(GetScatterNDGradient)
+DECLARE_GRADIENT_BUILDER(GetATenOpGradient)
 
 DECLARE_GRADIENT_BUILDER(GetExternalGradient)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.h:82]: I must be missing something... 
If ATenOp has gradientbuilder defined in python, it should take "GetExternalGradient" path. 
Only gradient-undefined ATenOp should take GetATenOpGradient path. 

I am missing the piece of code that is handling this logic.

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 5]
```diff
@@ -1714,6 +1714,26 @@ IMPLEMENT_GRADIENT_BUILDER(GetExternalGradient) {
   return result;
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetATenOpGradient) {
+  size_t num_inputs = GetSrcNodeInputSize();
+  std::vector<int64_t> requires_grad;
+  std::vector<ArgDef> output_args;
+  for (size_t i = 0; i < num_inputs; i++) {
+    if (IsGradientRequiredForSrcNodeInput(i)) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1722]: I realized that we are missing a mechanism to describe the differentiability for generic ATenOp. For regular ops, we use STOP_GRADIENT_EDGES to indicate which input is not differentiable. 

For ATenOp, we may introduce an attribute called "differentiable_inputs" to indicate which input is differentiable. 

Notice that this is different from the "requires_grad" attribute you are adding.

File: orttraining/orttraining/core/graph/gradient_builder_base.cc [hunk 6]
```diff
@@ -405,5 +405,22 @@ AttributeProto GradientBuilderBase::AttributeDefinitionToAttributeProto(
   return attr_proto;
 }
 
+// NOTE THAT we are changing the ATenOp forward node. If any of the input requires grad, we need to:
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder_base.cc:408]: Op's attribute is designed to hold **static** characteristic for a node's computation. 

I think we are breaking this assumption when manipulate the value of attribute for indicating which inputs needs gradient.

Comment (SherlockNoMad) [reply orttraining/orttraining/core/graph/gradient_builder_base.cc:408]: let's discuss offline on finding another way to pass this info.

File: orttraining/orttraining/core/graph/gradient_builder_base.cc (hunk 3 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder_base.cc:412]: The problem is we are using graph_ and node_ inside this function, and they are private in GradientBuilderBase. We need to move them from private to protected if we want to do this.

File: orttraining/orttraining/core/graph/gradient_builder.h (hunk 4 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.h:82]: The logic is at the beginning of GetGradientForOp in gradient_builder_registry.cc. Actually we will check external gradient for all node, the key for each node is domain::op_type (domain::op_type::aten_name::aten_overload_name for ATenOp), if the key is registered, external gradient will be used, otherwide, it will go to gradient builder by Op type.
